### PR TITLE
Fix product allocation check in dashboard

### LINKED ISSUE #577: Allocated line items should be computed while changing line quantity
A user makes an order for 5 items of the same product. 
Now we have this stock situation:  Quantity: 6,  Allocated:5

Go to the dashboard order page.
If we try to change the line quantity to 2-6 inside the dashboard we'll get an error: "Only 1 remaining in stock.".
The allocated line items should be added to the available stock.

### PR DESCRIPTION
**Pull request by krzysztofwolski**
Fix #577

### CODE REVIEW

**Comment by patrys** [comment]
Is love to have a test case! 😉

**Comment by krzysztofwolski** [comment]
Love delivered

File: tests/test_order.py [hunk 1]
```diff
@@ -35,3 +36,42 @@ def test_stock_allocation(billing_address, product_in_stock):
...
     stock = order_line.stock
     assert stock.quantity_allocated == 2
+
+
+def test_dashboard_change_quantity_form(billing_address, product_in_stock):
+    start_quantity = 1
+    variant = product_in_stock.variants.get()
+    cart = Cart()
+    cart.save()
+    cart.add(variant, quantity=start_quantity)
```
Comment (patrys) [inline tests/test_order.py]: Please use an existing fixture for that.

File: tests/test_order.py [hunk 2]
```diff
@@ -35,3 +36,42 @@ def test_stock_allocation(billing_address, product_in_stock):
...
+    order = models.Order.objects.create(billing_address=billing_address)
+    group = models.DeliveryGroup.objects.create(order=order)
+    group.add_items_from_partition(cart)
```
Comment (patrys) [inline tests/test_order.py]: Consider moving this to a fixture (`order`).

File: tests/test_order.py [hunk 3]
```diff
@@ -35,3 +36,42 @@ def test_stock_allocation(billing_address, product_in_stock):
...
+    order_line = group.items.get()
+    expected_quantity = order_line.stock.quantity_allocated
+
+    # Check available quantity validation
+    form = ChangeQuantityForm({'quantity': 9999},
+                              instance=order_line)
+    assert not form.is_valid()
+    assert group.items.get().stock.quantity_allocated == expected_quantity
```
Comment (patrys) [inline tests/test_order.py]: Not sure making `expected_quantity` a variable makes this easier to read.

File: tests/test_order.py [hunk 4]
```diff
@@ -35,3 +36,42 @@ def test_stock_allocation(billing_address, product_in_stock):
...
+    # Save same quantity
+    form = ChangeQuantityForm({'quantity': start_quantity},
```
Comment (patrys) [inline tests/test_order.py]: Same for `start_quantity`, use a number and assert on the number.

File: tests/test_order.py [hunk 5]
```diff
@@ -35,3 +36,42 @@ def test_stock_allocation(billing_address, product_in_stock):
...
+                              instance=order_line)
+    assert form.is_valid()
+    form.save()
+    assert group.items.get().stock.quantity_allocated == expected_quantity
+    # Increase quantity
+    expected_quantity += 1
```
Comment (patrys) [inline tests/test_order.py]: This requires me to backtrack and trace the control flow to determine what value is expected.

File: tests/test_order.py [hunk 6]
```diff
@@ -35,3 +36,42 @@ def test_stock_allocation(billing_address, product_in_stock):
...
+    form = ChangeQuantityForm({'quantity': start_quantity + 1},
+                              instance=order_line)
+    assert form.is_valid()
+    form.save()
+    assert group.items.get().stock.quantity_allocated == expected_quantity
+    # Decrease quantity
+    expected_quantity -= 1
```
Comment (patrys) [inline tests/test_order.py]: Same here.

**Comment by krzysztofwolski** [comment]
@patrys Created order fixture and ditched using variable for checking values as you suggested.

**Comment by krzysztofwolski** [comment]
@patrys Fixed

**Pull request merged**